### PR TITLE
Enable cpu_topology on s390x

### PR DESCRIPTION
**Pull request by fbq815**
CPU: updated sockets info, and enable exsisted test case on s390x

ID: 1925

Signed-off-by: Boqiao Fu <bfu@redhat.com>

### CODE REVIEW

**Review by smitterl — APPROVED** [review]
LGTM Thank you!

**Comment by fbq815** [comment]
Test result on RHEL9 guest:
RESULTS    : PASS 10 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0

Test result on RHEL8 guest:
failed by some product issue

@PaulYuuu @smitterl please help me review this patch

**Comment by fbq815** [comment]
Test result:
RHEL8:
 (2/2) io-github-autotest-qemu.cpu_topology_test.s390-virtio: STARTED
 (2/2) io-github-autotest-qemu.cpu_topology_test.s390-virtio: PASS (24.29 s)

RHEL9:
(1/1) io-github-autotest-qemu.cpu_topology_test.s390-virtio: STARTED
 (1/1) io-github-autotest-qemu.cpu_topology_test.s390-virtio: PASS (67.67 s)

@smitterl please help me review this again, I've updated the way to check the CPU topology since the issue on RHEL8, do you think this is needed? or it's better I remove the check point on RHEL8?

File: qemu/tests/cfg/cpu_topology_test.cfg [hunk 1]
```diff
@@ -13,3 +13,12 @@
...
             # Check PPTT table mapping by looking at core_id
             check_core_id_cmd = 'cat /sys/devices/system/cpu/cpu%s/topology/core_id'
+        s390x:
+            no RHEL.6 RHEL.7
+            no Host_RHEL.m6 Host_RHEL.m7 Host_RHEL.m8
+            required_qemu = [8.2.0-1,)
+            RHEL8:
+                check_siblings_cmd_rhel8 = 'core_siblings_count=$(($(cat /sys/devices/system/cpu/cpu0/topology/core_siblings_list | cut -d'-' -f2) - $(cat /sys/devices/system/cpu/cpu0/topology/core_siblings_list | cut -d'-' -f1) + 1)); echo "$core_siblings_count"'
+            RHEL.9:
+                check_siblings_cmd = 'cat /proc/cpuinfo |grep siblings |uniq |cut -d ":" -f 2'
```
Comment (PaulYuuu) [inline qemu/tests/cfg/cpu_topology_test.cfg:23]: Can use the same variable name in RHEL8 and RHEL9.

File: qemu/tests/cfg/cpu_topology_test.cfg [hunk 2]
```diff
@@ -13,3 +13,12 @@
             required_qemu = [6.2.0,)
             # Check PPTT table mapping by looking at core_id
             check_core_id_cmd = 'cat /sys/devices/system/cpu/cpu%s/topology/core_id'
+        s390x:
+            no RHEL.6 RHEL.7
+            no Host_RHEL.m6 Host_RHEL.m7 Host_RHEL.m8
+            required_qemu = [8.2.0-1,)
```
Comment (PaulYuuu) [inline qemu/tests/cfg/cpu_topology_test.cfg:19]:
```suggestion
            required_qemu = [8.2.0,)
```

File: provider/cpu_utils.py [hunk 3]
```diff
@@ -68,10 +69,14 @@ def check_if_vm_vcpu_topology_match(session, os_type, cpuinfo, test, devices=Non
     :param devices: qcontainer.DevContainer Object
     :return: True if guest topology is same as we expected
     """
+    host_arch = arch.ARCH
     if os_type == "linux":
         out = session.cmd_output_safe("lscpu")
         cpu_info = dict(re.findall(r"([A-Z].+):\s+(.+)", out, re.M))
-        sockets = int(cpu_info["Socket(s)"])
+        if host_arch == 's390x':
```
Comment (PaulYuuu) [inline provider/cpu_utils.py:76]: Depends on guest, not host.

File: qemu/tests/cpu_topology_test.py [hunk 4]
```diff
@@ -41,6 +41,8 @@ def check(p_name, exp, check_cmd):
         vcpu_threads_list = [1, 2, 4, 8]
     if 'arm64' in params['machine_type']:
         vcpu_threads_list = [1]
+    if 's390' in params['machine_type']:
+        vcpu_threads_list = [1]
```
Comment (PaulYuuu) [inline qemu/tests/cpu_topology_test.py:45]: Can merge into one.

File: qemu/tests/cfg/cpu_topology_test.cfg (hunk 1 above)
Comment (fbq815) [reply qemu/tests/cfg/cpu_topology_test.cfg:23]: updated

File: qemu/tests/cfg/cpu_topology_test.cfg (hunk 2 above)
Comment (fbq815) [reply qemu/tests/cfg/cpu_topology_test.cfg:19]: updated

File: qemu/tests/cpu_topology_test.py (hunk 4 above)
Comment (fbq815) [reply qemu/tests/cpu_topology_test.py:45]: updated

**Comment by fbq815** [comment]
Test result:
s390x.io-github-autotest-qemu.cpu_topology_test.check_siblings_core_id_cmd.s390-virtio: PASS

@PaulYuuu please help me review this patch

File: qemu/tests/cfg/cpu_topology_test.cfg [hunk 5]
```diff
@@ -13,3 +13,15 @@
...
             check_core_id_cmd = 'cat /sys/devices/system/cpu/cpu%s/topology/core_id'
+        s390x:
+            no RHEL.6 RHEL.7
+            no Host_RHEL.m6 Host_RHEL.m7 Host_RHEL.m8
+            required_qemu = [8.2.0,)
+    variants:
+        - check_siblings_core_id_cmd:
+            only s390x
+            RHEL.8:
+                check_siblings_cmd = 'core_siblings_count=$(($(cat /sys/devices/system/cpu/cpu0/topology/core_siblings_list | cut -d'-' -f2) - $(cat /sys/devices/system/cpu/cpu0/topology/core_siblings_list | cut -d'-' -f1) + 1)); echo "$core_siblings_count"'
```
Comment (PaulYuuu) [inline qemu/tests/cfg/cpu_topology_test.cfg:24]:
```suggestion
                check_siblings_cmd = 'core_siblings_count=$(($(awk -F'-' '{print $2 - $1 + 1}' /sys/devices/system/cpu/cpu0/topology/core_siblings_list))); echo "$core_siblings_count"'
```

File: qemu/tests/cfg/cpu_topology_test.cfg [hunk 6]
```diff
@@ -13,3 +13,15 @@
             required_qemu = [6.2.0,)
             # Check PPTT table mapping by looking at core_id
             check_core_id_cmd = 'cat /sys/devices/system/cpu/cpu%s/topology/core_id'
+        s390x:
+            no RHEL.6 RHEL.7
+            no Host_RHEL.m6 Host_RHEL.m7 Host_RHEL.m8
+            required_qemu = [8.2.0,)
+    variants:
+        - check_siblings_core_id_cmd:
```
Comment (PaulYuuu) [inline qemu/tests/cfg/cpu_topology_test.cfg:21]: There is only 1 variant, so other architectures cannot find the test case, as the only one variant is defined with `only s390x`.

Comment (fbq815) [reply qemu/tests/cfg/cpu_topology_test.cfg:21]: updated

File: qemu/tests/cfg/cpu_topology_test.cfg (hunk 5 above)
Comment (fbq815) [reply qemu/tests/cfg/cpu_topology_test.cfg]: updated

File: qemu/tests/cfg/cpu_topology_test.cfg (hunk 6 above)
Comment (PaulYuuu) [inline qemu/tests/cfg/cpu_topology_test.cfg:21]: So this changed the test case name, I would include @nanliu-r @zhenyzha @maxujun, to consider this change.

File: qemu/tests/cfg/cpu_topology_test.cfg [hunk 7]
```diff
@@ -13,3 +13,15 @@
...
+            s390x:
+                RHEL.8:
+                    check_siblings_cmd = 'awk -F'-' '{print $2 - $1 + 1}' /sys/devices/system/cpu/cpu0/topology/core_siblings_list'
+                RHEL.9:
+                    check_siblings_cmd = 'cat /proc/cpuinfo |grep siblings |uniq |cut -d ":" -f 2'
+                check_core_id_cmd = 'cat /sys/devices/system/cpu/cpu%s/topology/core_id'
```
Comment (zhenyzha) [inline qemu/tests/cfg/cpu_topology_test.cfg:27]: `check_core_id_cmd = 'cat /sys/devices/system/cpu/cpu%s/topology/core_id'`
This configuration is also in s390. Is it set for rhel7 guest?

File: qemu/tests/cfg/cpu_topology_test.cfg [hunk 8]
```diff
@@ -13,3 +13,15 @@
...
             check_core_id_cmd = 'cat /sys/devices/system/cpu/cpu%s/topology/core_id'
+        s390x:
+            no RHEL.6 RHEL.7
+            no Host_RHEL.m6 Host_RHEL.m7 Host_RHEL.m8
+            required_qemu = [8.2.0,)
+    variants:
+        - check_siblings_core_id_cmd:
+            s390x:
+                RHEL.8:
+                    check_siblings_cmd = 'awk -F'-' '{print $2 - $1 + 1}' /sys/devices/system/cpu/cpu0/topology/core_siblings_list'
```
Comment (PaulYuuu) [inline qemu/tests/cfg/cpu_topology_test.cfg:24]:
```suggestion
                    check_siblings_cmd = "awk -F'-' '{print $2 - $1 + 1}' /sys/devices/system/cpu/cpu0/topology/core_siblings_list"
```
Make it more human-readable.

File: qemu/tests/cfg/cpu_topology_test.cfg (hunk 6 above)
Comment (fbq815) [reply qemu/tests/cfg/cpu_topology_test.cfg:21]: I'll add other variant later but with other patch

File: qemu/tests/cfg/cpu_topology_test.cfg (hunk 8 above)
Comment (fbq815) [reply qemu/tests/cfg/cpu_topology_test.cfg]: updated

File: qemu/tests/cfg/cpu_topology_test.cfg (hunk 6 above)
Comment (fbq815) [reply qemu/tests/cfg/cpu_topology_test.cfg]: after discussion, the original test cases name would expand with "default"
@nanliu-r @zhenyzha @maxujun please help me review if this works on your platform

**Comment by fbq815** [comment]
Test result on s390x:
cpu_topology_test.siblings_core.s390-virtio: PASS (20.74 s)

**Comment by nanliu-r** [comment]
Test PASS on x86.

**Review by zhenyzha — APPROVED** [review]
Pass on aarch64 platform

LGTM.Acked-by: Zhenyu Zhang <zhenyzha@redhat.com>

**Review by nanliu-r — APPROVED** [review]

File: qemu/tests/cfg/cpu_topology_test.cfg [hunk 9]
```diff
@@ -13,3 +13,17 @@
...
+    variants:
+        - default:
+            no s390x
+        - siblings_core:
+            only s390x
+                RHEL.8:
+                    check_siblings_cmd = "awk -F'-' '{print $2 - $1 + 1}' /sys/devices/system/cpu/cpu0/topology/core_siblings_list"
+                RHEL.9:
+                    check_siblings_cmd = 'cat /proc/cpuinfo |grep siblings |uniq |cut -d ":" -f 2'
+                check_core_id_cmd = 'cat /sys/devices/system/cpu/cpu%s/topology/core_id'
```
Comment (PaulYuuu) [inline qemu/tests/cfg/cpu_topology_test.cfg:29]:
```suggestion
            RHEL.8:
                check_siblings_cmd = "awk -F'-' '{print $2 - $1 + 1}' /sys/devices/system/cpu/cpu0/topology/core_siblings_list"
            RHEL.9:
                check_siblings_cmd = 'cat /proc/cpuinfo |grep siblings |uniq |cut -d ":" -f 2'
            check_core_id_cmd = 'cat /sys/devices/system/cpu/cpu%s/topology/core_id'
```

Comment (fbq815) [reply qemu/tests/cfg/cpu_topology_test.cfg]: updated

**Review by PaulYuuu — APPROVED** [review]

**Pull request merged**